Update from 05/15/2021 for Assets Documentation (#497)
osisoft/OCS-Docs: content/api-reference/assets/asset-search-api.md

`string queryString`  
The asset query string. Search strings are not case-sensitive.

`string filterString`  

> The asset filter string. Filter strings are not case-sensitive. At this moment, only asset AssetTypeName and asset metadata Name to Value can be filtered on.

Reviewer: ```suggestion
String used to filter the asset search results. Filter strings are not case sensitive. The strings on which you can filter results are limited to the asset AssetTypeName and the asset metadata Name to Value properties.
```
Is this correct? What does "metadata Name to Value" mean?

Author: There is probably a better way to say this, but what I meant was that we can do a filter like : filter[location]=Earth the name is "location" and the value is "Earth". Open for suggestions

Reviewer: Please look at my edits. Is this correct?

Author: Makes sense to me. Thanks

[optional] `[id|name] [asc|desc|] orderby`  
An optional parameter which returns assets ordered either by the asset `Id` or the asset `name`. Specify asc or desc to return the results in ascending or descending order. If not specified, the default is ascending order.

[optional] `int pageSize`   

> Used in conjunction with `maxPages` for search result pagination. This parameter defines the max number of assets that should be returned per page. 

Reviewer: ```suggestion
Used with `maxPages` to configure the search results. This parameter defines the maximum number of assets that are returned on each page. 
```

Author: Thanks.

Reviewer: This change has been committed.

> Used in conjunction with `maxPages` for search result pagination. This parameter defines the max number of assets that should be returned per page. 

Reviewer: ```suggestion
Used with `maxPages` to configure the search results. This parameter defines the maximum number of assets that are returned on each page. 
```

[optional] `int maxPages`   

> Used in conjunction with `maxPages` for search result pagination. This parameter defines the max number of pages.

Reviewer: ```suggestion
Used with `pageSize` for search result pagination. This parameter defines the max number of pages.
```

Author: If the number of results if larger than the maxPages, then the search will be limited to the maxPages, that is not all search results will be returned.

Reviewer: Should we say something like: Search results that exceed `maxPages` are not returned. Or is that obvious.

Do you have to specify both pageSize and maxPages? If yes, what happens if you only specify one of the parameters.

Author: The user always has to specify both parameters. I think that it's good to say " Search results that exceed maxPages are not returned. " since it might create confusion if not.

> Used in conjunction with `maxPages` for search result pagination. This parameter defines the max number of pages.

Reviewer: ```suggestion
Must be used with `pageSize` to configure the pages of the search results. This parameter defines the maximum number of pages. Search results that exceed maxPages are not returned.
```

> Used in conjunction with `maxPages` for search result pagination. This parameter defines the max number of pages.

Reviewer: ```suggestion
Must be used with `pageSize` to configure the pages of the search results. This parameter defines the maximum number of pages. Search results that exceed maxPages are not returned.
```

Author: Hmm, says it's outdated, do you need to merge from main first?

Reviewer: This change has been committed.

[optional] `string continuationToken`   

> An optional parameter used for search pagination to retrieve a given page in the paged search results.

Reviewer: ```suggestion
Retrieves a specific page in the paged search results.
```
I'm assuming it's obvious how they do this?

Author: Ah yes! I commented below, but thanks for catching this!

`string queryString`  
The asset query string. Search strings are not case-sensitive.

`string filterString`  

> The asset filter string. Filter strings are not case-sensitive. At this moment, only asset AssetTypeName and asset metadata Name to Value can be filtered on.

Reviewer: ```suggestion
String used to filter the asset search results. Filter strings are not case sensitive. The strings on which you can filter results are limited to the asset AssetTypeName property and the asset metadata properties (using the syntax filter[*property_name*]=*property_value*).
```
Replacing with the same text used above.

Author: Looks good

[optional] `[id|name] [asc|desc|] orderby`  
An optional parameter which returns assets ordered either by the asset `Id` or the asset `name`. Specify asc or desc to return the results in ascending or descending order. If not specified, the default is ascending order.

> 

Reviewer: Once we make the changes above to the same parameters, I'll duplicate the changes here.

Author: Thanks!

[optional] `int pageSize`   

> Used in conjunction with `maxPages` for search result pagination. This parameter defines the max number of assets that should be returned per page. 

Reviewer: ```suggestion
Must be used with `maxPages` to configure the pages of the search results. This parameter defines the maximum number of assets that are returned on each page. 
```

Author: Looks good

[optional] `int maxPages`   

> Used in conjunction with `maxPages` for search result pagination. This parameter defines the max number of pages.

Reviewer: ```suggestion
Must be used with `pageSize` to configure the pages of the search results. This parameter defines the maximum number of pages. Search results that exceed maxPages are not returned.
```

Author: Looks good

[optional] `string continuationToken`   

> An optional parameter used for search pagination to retrieve a given page in the paged search results.

Reviewer: ```suggestion
Retrieves a specific page of the search results. 
```

### Special characters in search queries

Add the backslash escape character ( \ ) before any special characters in search queries. The following special characters require an escape character:   " |  /  *  \  ( )  : 

The following are examples of using the escape character in query strings.

| Example Field Value                    | Query String                               |
| -------------------------------------- | ------------------------------------------ |
| Austin\Dallas\Fort Worth               | Austin\\\Dallas\\\Fort Worth               |
| 1:100                                  | 1\\:100                                    |

### Examples of asset filter strings

Filter strings are not case sensitive. Numeric types must be passed as strings according to English locale. For example, a double of 1.1 must be sent as "1.1".

| Query String                                 | Description                                                  |
| -------------------------------------------- | ------------------------------------------------------------ |

> | filter[location]=Earth                       | Returns all assets that contains a metadata name = **location** and value is **Earth**. |

Reviewer: ```suggestion
| filter[location]=Earth                       | Returns all assets that contain a metadata name = **location** and value is **Earth**. |
```

Author: looks good.

> | filter[location]=Earth&filter[device]=tracer | Returns all assets that contains a metadata name = **location** and value is **Earth** and also contains a metadata with name = **device** and value is **tracer**. |

Reviewer: ```suggestion
| filter[location]=Earth&filter[device]=tracer | Returns all assets that contain a metadata name = **location** and value is **Earth** and also contains a metadata with name = **device** and value is **tracer**. |
```

Author: Looks good.

> ## `Asset Search Pagination` 

Reviewer: ```suggestion
## `Configure Asset Search Results` 
```
I just noticed this. As I mentioned in another comment, "Pagination" refers to page numbering which is not what we are doing here. You could also say "Asset Search Results by Pages"  or "Configure Asset Search Results Display."

Author: I think "Asset Results by Pages" makes sense since we can get paged results even outside of search. Perhaps because of that, maybe it doesn't belong in this section, but I think it's fine for now.

> Asset search results can be returned in pages through `pageCount` ,`maxPages`, and `continuationToken` query parameters.

Author: Correction 'pagecount' should be 'pageSize'. Apologize for this.

Reviewer: This change has been committed.

> Asset search results can be returned in pages through `pageCount` ,`maxPages`, and `continuationToken` query parameters.

Reviewer: ```suggestion
Asset search results can be returned in pages through `pageSize` ,`maxPages`, and `continuationToken` query parameters.
```

> Asset search results can be returned in pages through `pageCount` ,`maxPages`, and `continuationToken` query parameters.

Reviewer: ```suggestion
Asset search results can be returned in pages using the `pageSize` ,`maxPages`, and `continuationToken` query parameters.
```

Author: Thanks for updating.

> Asset search results can be returned in pages through `pageCount` ,`maxPages`, and `continuationToken` query parameters.

Reviewer: ```suggestion
Asset search results can be returned in pages through `pageSize` ,`maxPages`, and `continuationToken` query parameters.
```

Author: Thanks for updating.

Reviewer: This change has been committed

> When you want asset search results returned in pages, you first need to make a call using the `pageCount` and `maxPages` parameters. For example:

Reviewer: ```suggestion
To return search results in pages, first make a call using the `pageCount` and `maxPages` parameters. For example:
```

Author: pageCount should be pageSize.

Reviewer: This change has been committed.

> `GET search/assets?pageCount=10&maxPages=100` will request search results with a maximum of 10 assets per page and a maximum of 100 pages. 

Reviewer: ```suggestion
`GET search/assets?pageCount=10&maxPages=100` returns search results with a maximum of 10 assets per page and a maximum of 100 pages. 
```

Author: looks good besides that pageCount should be pageSize

> `GET search/assets?pageCount=10&maxPages=100` will request search results with a maximum of 10 assets per page and a maximum of 100 pages. 

Reviewer: ```suggestion
`GET search/assets?pageCount=10&maxPages=100` returns search results with a maximum of 10 assets per page and a maximum of 100 pages. 
```

Author: Looks good.

Reviewer: Change committed

> The HTTP header of this response of this query will include a "Link" section to which will define either first, next, previous, or last pages. These links can be used directly to follow the pages. For example, the links section may contain:

Reviewer: ```suggestion
The HTTP header of this response of this query will include a "Link" section which will define links to the first, next, previous, and last pages. These links can be used directly to follow the pages. For example, the links section may contain:
```

Author: looks good.

```json 

> <https://{clusterName}/api/v1-preview/tenants/{tenantId}/namespaces/{NamespaceId}/Assets?continuationToken={continuationToken}>; rel="next", 

Reviewer: ```suggestion
<https://{cluster}/api/v1-preview/tenants/{tenant}/namespaces/{Namespace}/Assets?continuationToken={continuationToken}>; rel="next", 
```

<https://{clusterName}/api/v1-preview/tenants/{tenantId}/namespaces/{NamespaceId}/Assets?continuationToken={continuationToken}>; rel="previous",
<https://{clusterName}/api/v1-preview/tenants/{tenantId}/namespaces/{NamespaceId}/Assets?continuationToken={continuationToken}>; rel="first",
<https://{clusterName}/api/v1-preview/tenants/{tenantId}/namespaces/{NamespaceId}/Assets?continuationToken={continuationToken}>; rel="last"

> ```

Reviewer: What is the meaning of red text?

Author: hmm, I was using an editor, the text shouldn't be any different than normal JSON.

> These links can simply be followed to get the next, previous, first, and last page. Pagination search results timeout in 2 minutes if no pages are requested.

Reviewer: ```suggestion
These links can be followed to get the next, previous, first, and last page. Pagination search results time out in 2 minutes if no pages are requested.
```

Author: looks good.

## `Asset Faceted Search` 

> Asset faceted search allows searching by asset facets. Asset facets are not case sensitive. At this moment, only asset metadata is facetable.

Reviewer: ```suggestion
Asset faceted search allows for searching by asset facets. Asset facets are not case sensitive.  Only asset metadata can be used in asset faceted searches.
```

Author: Changes look good.

### Request 

```text 
GET api/v1-preview/Tenants/{tenantId}/Namespaces/{namespaceId}/search/Assets/facets?name={category}&count={count}
```

### Parameters  

`string tenantId`  
The tenant identifier

`string namespaceId`  
The namespace identifier

`string category`  

> The asset metadata name you want to get all facets of.[optional] 

Reviewer: ```suggestion
[optional] The name of the asset metadata for which you want to retrieve the facet values.
```

Author: The metadata name is not optional. Just removing [optional] if front is good. Thanks!

> The asset metadata name you want to get all facets of.[optional] 

Reviewer: Pls explain

Author: Let's talk about this today so that I can clarify.

[optional] `int count`   
An optional parameter, between 1 and 1000 (inclusive), representing the maximum number of retrieved assets. If not specified, the default is 100.

### Response 

A list of asset Ids and their matched fields.

| Status Code     | Body Type        | Description                                                  |
| --------------- | ---------------- | ------------------------------------------------------------ |

> | 200 OK          | Asset facet list | A list of assets facet matching the search criteria. An empty array is returned if there are no matching asset metadata with the supplied category. |

Reviewer: ```suggestion
| 200 OK          | Asset facet list | A list of asset facet values that match the search criteria. An empty array is returned if there are no matching asset metadata with the supplied category. |
```

Author: looks good.

| 204 OK          | empty            | No assets match your search criteria.                        |
| 400 Bad Request | error            | The request is not valid. See the response body for additional details. |

### Example 

If you have the following assets in your system:

| Assets in System                                             |
| ------------------------------------------------------------ |
| "Id": "CaliforniaAsset",    "Metadata": [      { "Name": "Manufacturer", "Value": "GE", "SdsTypeCode": "String" },      { "Name": "Location", "Value": "California", "SdsTypeCode": "String" },    ] |
| "Id": "CaliforniaAsset1",    "Metadata": [      { "Name": "Manufacturer", "Value": "Rockwell", "SdsTypeCode": "String" },      { "Name": "Location", "Value": "California", "SdsTypeCode": "String" },    ] |
| "Id": "PhillyAsset",    "Metadata": [      { "Name": "Manufacturer", "Value": "GE", "SdsTypeCode": "String" },      { "Name": "Location", "Value": "Philly", "SdsTypeCode": "String" },    ] |

> Performing a `GET search/assets/facets?Name=Location` will return the following response. 

Reviewer: ```suggestion
Performing a `GET search/assets/facets?Name=Location` returns the following response. 
```

Author: Thanks.
